### DISCUSSION TITLE
Map GS NPRN Params to CC numbers

### DISCUSSION THREAD

**Discussion by spessasus** [FluidSynth/fluidsynth#1406] [Ideas]
## The problem
Currently, there's no way to make fluidsynth react to GS NPRNs.

For example, take a listen to [this MIDI file being played on a GS device](https://youtu.be/0jsOL5Og-0Q). At around 18 seconds in, the trumpets can be heard with clear vibrato. Playing [it](https://github.com/AyHa1810/touhou-midi-collection/blob/main/1%20-%20Shooter%20Games/7%20-%20Perfect%20Cherry%20Blossom/14.%20Border%20of%20Life%20(ZUN).mid) on fluidsynth even with stgiga's HiDef (the SC-88Pro soundfont) does not result in any vibrato.

And that's fine, fluidsynth is not said to be GS compatible.

But, even as a soundfont creator, I *can't* make it GS compatible with modulators. You see, modulators only support CCs as sources, not NPRNs. This is why I propose the following:

## The solution
In this [SC8850 owner's manual](https://cdn.roland.com/assets/media/pdf/SC-8850_OM.pdf#page=227&zoom=133,-64,835) you can see the CC numbers on page 227 and the correspoding NPRNs on page 228.

It would be great if fluidsynth *mapped* those NPRNs to their respective controllers. That is:
- NPRN 01H 08H to CC 76 (vibrato rate)
- NPRN 01H 09H to CC 77 (vibrato depth)
- NPRN 01H 0AH to CC 78 (vibrato delay)

`note: numbers with "H" at the end indicate hex`

And by mapping i mean directly copying the data entry value to the CC:
- received an NRPN LSB: 01H
- received an NRPN MSB: 08H (this is vibrato rate)
- received a data entry MSB: 10H: this is effectively interpreted as CC change: CC 76 is set to 10H

**Note: this would only apply if the synth is in the `gs` bank select mode (I.E. after receiving a GS sysex)

## Rationale
- does not mess up other systems (the change only applies after a GS sysEx)
- does not add any default modulators or anything like that
- allows the soundfont creators to implement support for the said GS NRPNs
- does not break any existing soundfonts

What do you think?

**Comment by derselbst** [comment]
I don't have a strong opinion here. Converting the NRPNs would obviously clobber the respective CCs. Since that would be limited to gs bank select mode it would be fine I guess. The soundfont would still need to implement modulators to achieve the desired effect.

On the other hand, one could convert the GS NRPNs to SF2 NRPNs to directly influence the respective destination generators. Have you considered that?

**Reply by spessasus** [reply]
I don't think you understood me correctly. We want to avoid modifying the song. If we do, then I would just convert NPRNs into CC changes. What I want to achieve is making the synth respond to given NRPNs using modulators.

But I just thought of an even better idea, making it an extension to the INFO chunk. Give me a moment, I'll write the proposal and link it here.

**Reply by derselbst** [reply]
I did understand it as on-the-fly-NRPN-to-CC conversion by fluidsynth's internal logic. But if the MIDI file sets e.g. CC76, that CC would be clobbered by the NRPN.

"making the synth respond to given NRPNs using modulators" is only feasible, if there is a reliable and general source of how the implementation of such modulators would look like, e.g. what amount, transform, etc.

**Reply by spessasus** [reply]
Yes, precisely. Then I can define the behavior using modulators. by default, these NPRNs would not do anything (since there are no default modulators that use these CCs as sources). The GS compatible soundfont then can set how it thinks the vibrato should behave.

This implementation *would not* make fluid respond to these by default. It just *gives the ability* to map these NRPNs to modulators by the soundfont's creator.

And about clobbering, I think you didn't consider this:
these CCs serve *the same exact purpose* as NRPNs in GS. If MIDI uses the NPRNs, it probably doesn't use the CCs. and even then, it would also overwrite the vibrato on the actual GS device. For example, BASSMIDI documentation:
> MIDI_EVENT_VIBRATO_RATE	Set the vibrato rate (MIDI controller 76, NRPN 108h).
param : vibrato rate (0-127, 0=-64, 64=normal, 127=+63). 

As you can see, both of these are interpreted identically.

**Reply by derselbst** [reply]
Ok. I'd be open to give it a try.
